cpp/tree.hpp

```diff
@@ -0,0 +1,262 @@
+#pragma once
+
+/**
+ * @file tree.hpp
+ * @brief 木の汎用テンプレート
+*/
+
+#include <stack>
+
+#include "graph.hpp"
+
+template <typename>
+struct RootedTree;
+template <typename>
+struct DoublingClimbTree;
+
+/**
+ * @brief 根なし木
+ * 
+ * @tparam T = int 辺の重み
+ * 
+ * Graph<T>を継承し、すべて無向辺で表す
+*/
+template <typename T = int>
+struct Tree : public Graph<T> {
+    /**
+     * @brief コンストラクタ
+     * 
+     * @param n ノード数
+     */
+    Tree(int n = 0) : Graph<T>(n) {}
+
+    /**
+     * @brief 辺の情報を標準入力から受け取って追加する
+     * @param padding  = -1 頂点番号を入力からいくつずらすか
+     * @param weighted = false 辺の重みが入力されるか
+     */
+    void read(int padding = -1, bool weighted = false) {
+        Graph<T>::read(this->n - 1, padding, weighted, false);
+    }
+
+    /**
+     * @brief 根付き木にする
+     * 
+     * @param root 根
+     * @return RootedTree<T> 根付き木のオブジェクト
+     */
+    [[nodiscard]]
+    RootedTree<T> set_root(int root) const {
+        return RootedTree<T>(*this, root);
+    }
+    /**
+     * @brief 根付き木にして返す
+     * 
+     * @param root 根
+     * @return RootedTree<T> 根付き木のオブジェクト
+     * 
+     * set_root()に比べてコピーコストがかからないが、
+     * ムーブされるため、このオブジェクトは使用不可になる
+     * 複数種類の根を持った根付き木を得たい場合、set_root()を使う
+     */
+    [[nodiscard]]
+    RootedTree<T> set_root_move(int root) {
```

Comment:
変更後のメンバ関数の呼び方が分からないような人を対象にしてるなら、そのままでもいいかも。

```suggestion
    RootedTree<T> set_root(int root) && {
```